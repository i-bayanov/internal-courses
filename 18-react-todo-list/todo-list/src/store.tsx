import { configureStore } from '@reduxjs/toolkit';
import { IAction } from './interfaces-and-types';
import todoReducer from './reducer';

const store = configureStore({ reducer: todoReducer });

const dispatchAndSave = (action: IAction) => {
  store.dispatch(action);
  localStorage.setItem('todos', JSON.stringify(store.getState()));
};

type RootState = ReturnType<typeof store.getState>;

export { store, dispatchAndSave, RootState };
