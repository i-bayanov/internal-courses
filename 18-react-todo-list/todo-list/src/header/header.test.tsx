import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';

import Header from './header';
import * as store from '../store';

let container: Element | null = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container!);
  container!.remove();
  container = null;
});

describe('Header component', () => {
  it('should renders correctly', () => {
    render(<Header />, container!);
    expect(container).toMatchSnapshot();
  });

  it('should change document title when new too added', (done) => {
    jest.useFakeTimers();
    const spy = jest.spyOn(store, 'dispatchAndSave');
    const newTitle = '***********************************Added new TODO: New Todo***********************************';

    render(<Header />, container!);
    const input = container?.querySelector('.new-todo') as HTMLInputElement;
    input.value = 'New Todo';
    jest.setSystemTime(1652355429278);
    input.dispatchEvent(new KeyboardEvent('keydown', {
      key: 'Enter',
      bubbles: true,
    }));
    expect(spy).toHaveBeenCalledWith({
      type: 'todo/add',
      payload: {
        todo: {
          id: String(1652355429278),
          title: 'New Todo',
          completed: false,
        },
      },
    });
    jest.useRealTimers();
    jest.restoreAllMocks();
    setTimeout(() => {
      expect(document.title).toBe(newTitle);
      done();
    });
  });
});
