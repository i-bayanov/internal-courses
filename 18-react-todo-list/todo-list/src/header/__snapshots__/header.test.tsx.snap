// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should renders correctly 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      todos
    </h1>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
    />
  </header>
</div>
`;
